pkg/csi/service/wcpguest/controller.go

```diff
@@ -566,59 +566,83 @@ func (c *controller) ControllerExpandVolume(ctx context.Context, req *csi.Contro
-		pvc, err = c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Update(ctx, pvcClone, metav1.UpdateOptions{})
+	waitForSvPvcCondition := true
+	gcPvcRequestSize := resource.NewQuantity(volSizeBytes, resource.Format(resource.BinarySI))
+	svPvcRequestSize := svPVC.Spec.Resources.Requests[corev1.ResourceName(corev1.ResourceStorage)]
+	// Check if GC PVC request size is greater than SV PVC request size
+	if (gcPvcRequestSize).Cmp(svPvcRequestSize) == 1 {
+		// Update requested storage in SV PVC spec
+		svPvcClone := svPVC.DeepCopy()
+		svPvcClone.Spec.Resources.Requests[corev1.ResourceName(corev1.ResourceStorage)] = *gcPvcRequestSize
+
+		// Make an update call to SV API server
+		log.Infof("Increasing the size of supervisor PVC %s in namespace %s to %s", volumeID, c.supervisorNamespace, gcPvcRequestSize.String())
+		svPVC, err = c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Update(ctx, svPvcClone, metav1.UpdateOptions{})
 		if err != nil {
 			msg := fmt.Sprintf("failed to update supervisor PVC %q in %q namespace. Error: %+v", volumeID, c.supervisorNamespace, err)
 			log.Error(msg)
 			return nil, status.Error(codes.Internal, msg)
 		}
-	} else {
-		log.Debugf("Skipping resize call for supervisor PVC %s in namespace %s", volumeID, c.supervisorNamespace)
-	}
+	} else if (gcPvcRequestSize).Cmp(svPvcRequestSize) == 0 {
+		// GC PVC request size is equal to SV PVC request size
+		log.Infof("Skipping resize call for supervisor PVC %s in namespace %s", volumeID, c.supervisorNamespace)
 
-	// Wait for Supervisor PVC to change status to FilesystemResizePending
-	err = checkForSupervisorPVCCondition(ctx, c.supervisorClient, pvc,
-		corev1.PersistentVolumeClaimFileSystemResizePending, time.Duration(getResizeTimeoutInMin(ctx))*time.Minute)
-	if err != nil {
-		msg := fmt.Sprintf("failed to expand volume %s in namespace %s of supervisor cluster. Error: %+v", volumeID, c.supervisorNamespace, err)
+		// SV PVC is already in FileSystemResizePending condition -> SV PV has already been expanded to required size
+		if checkPVCCondition(ctx, svPVC, corev1.PersistentVolumeClaimFileSystemResizePending) {
+			waitForSvPvcCondition = false
+		} else {
+			// SV PVC is not in FileSystemResizePending condition and GC PVC request size is equal to SV PVC capacity -> SV PVC is already at required size
+			if (gcPvcRequestSize).Cmp(svPVC.Status.Capacity[corev1.ResourceName(corev1.ResourceStorage)]) == 0 {
+				waitForSvPvcCondition = false
+			}
+		}
+	} else {
+		// GC PVC request size is lesser than SV PVC request size
+		msg := fmt.Sprintf("the requested size of the Supervisor PVC %s in namespace %s is %s which is greater than the requested size of %s",
+			volumeID, c.supervisorNamespace, svPvcRequestSize.String(), gcPvcRequestSize.String())
 		log.Error(msg)
 		return nil, status.Error(codes.Internal, msg)
```

Comment:
How about changing the error code to `InvalidArgument` ?

Reply:
It is not an `InvalidArgument` right? Ideally that request size would go through in Vanilla flavor. It is just not compatible with the supervisor layer we have introduced underneath. IMO it should be internal itself.

Reply:
> Ideally that request size would go through in Vanilla flavor.

How? If the backend volume size is greater than the ExpandVolume request size, CNS will reject the request. IIUC, the system will enter this state if a user manually expands the SV-PVC and therefore the TKG PVC has no idea about it. That's an invalid operation

Reply:
Changed it